readme.md
# 番号

A Ruby library for converting between positive integers and Sino-Japanese numbers or text.

[![Build Status](https://travis-ci.org/kattrali/bangou.png?branch=master)](https://travis-ci.org/kattrali/bangou) [![Code Climate](https://codeclimate.com/badge.png)](https://codeclimate.com/github/kattrali/bangou)

## Usage

```
require 'bangou'


89.to_j
# => "八十九"

289.to_jtext
# => "二百八十九"

Bangou.integer_to_japanese_numerals(2305)
# => "二千三百五"

Bangou.integer_to_japanese_text(2305)
# => "にせんさんびゃくご"
```

## Installation

`gem install bangou`

## Running Tests

番号 uses bacon for testing, runnable using `rake spec`.

## Contributing

Pull requests with tests accepted!

## Current Limitations

Only works on numbers between 0 and 99,999,999 (otherwise throwing an OutOfRangeException).


changelog.md
# 1.1.1

 - Added better class documentation

# 1.1

 - Added extension on Integer to simplify calling the interface:
 
```
89.to_j
# => "八十九"

289.to_jtext
# => "二百八十九"
```

# 1.0

Initial release

spec/bangou_spec.rb
# encoding: UTF-8
$:.unshift(File.expand_path('../../lib',__FILE__))

require 'bangou.rb'

describe "Bangou" do
	describe "changes integers into Japanese numerals" do

		def compare integer, numeral
			Bangou.integer_to_japanese_numerals(integer).should.equal(numeral)
		end

		it "uses the correct bases for numbers" do
			compare(1,     "一")
			compare(10,    "十")
			compare(100,   "百")
			compare(1000,  "千")
			compare(10000, "万")
		end

		it "can parse digits in different places in a number, ignoring zeroes" do
			compare(7509,  "七千五百九")
			compare(80000, "八万")
			compare(80001, "八万一")
			compare(80020, "八万二十")
		end

    it "removes unneeded ones" do
      compare(81000, "八万千")
			compare(10100, "万百")
    end

		it "can handle bases above 10^4" do
			compare(100000,    "十万")
			compare(1000000,   "百万")
			compare(50000000,  "五千万")
			compare(22000000,  "二千二百万")
		end

    it "raises OutOfRangeException if given a number out of usable range" do
      should.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_numerals(100000000)
      }
      should.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_numerals(-1)
      }
      should.not.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_numerals(99999999)
      }
      should.not.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_numerals(0)
      }
    end
	end

	describe "changes integers into Japanese text" do
		def compare integer, text
			Bangou.integer_to_japanese_text(integer).should.equal(text)
		end

		it "ignores zeroes, unless the number is just zero" do
			compare(7509,  "ななせんごひゃくきゅう")
      compare(0,     "ゼロ")
			compare(80000, "はちまん")
			compare(80001, "はちまんいち")
			compare(80020, "はちまんにじゅう")
		end

    it "raises OutOfRangeException if given a number out of usable range" do
      should.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_text(100000000)
      }
      should.not.raise(Bangou::OutOfRangeException) {
        Bangou.integer_to_japanese_text(99999999)
      }
    end

		it "uses the correct transformation for 3, 6, and 8" do
			compare(3, "さん")
			compare(6, "ろく")
			compare(8, "はち")

			compare(30, "さんじゅう")
			compare(60, "ろくじゅう")
			compare(80, "はちじゅう")

			compare(300, "さんびゃく")
			compare(600, "ろっぴゃく")
			compare(800, "はっぴゃく")

			compare(3000, "さんぜん")
			compare(6000, "ろくせん")
			compare(8000, "はっせん")

			compare(30000, "さんまん")
			compare(60000, "ろくまん")
			compare(80000, "はちまん")

			compare(300000, "さんじゅうまん")
			compare(600000, "ろくじゅうまん")
			compare(800000, "はちじゅうまん")

			compare(3000000, "さんびゃくまん")
			compare(6000000, "ろっぴゃくまん")
			compare(8000000, "はっぴゃくまん")
		end
	end
end

spec/integer_spec.rb
# encoding: UTF-8
$:.unshift(File.expand_path('../../lib',__FILE__))

require 'bangou.rb'

describe "Integer additions" do
	it "adds Integer#to_j which is the kanji equivalent of a given integer" do
		89.to_j.should.equal "八十九"
		289.to_j.should.equal "二百八十九"
	end

	it "adds Integer#to_jtext which is the hiragana equivalent of a given integer" do
		89.to_jtext.should.equal "はちじゅうきゅう"
		289.to_jtext.should.equal "にひゃくはちじゅうきゅう"
	end
end

lib/bangou.rb
# encoding: UTF-8

require 'extensions/integer.rb'

class Bangou
  class OutOfRangeException < Exception; end

  # Convert an Integer into Japanese numerals
  #
  # Example:
  #   Bangou.integer_to_japanese_numerals(89)
  #   # => "八十九"
  #
  # @param [Integer] the number to convert
  # @return [String] The Japanese numeric equivalent in kanji
  def self.integer_to_japanese_numerals int
    convert_number(int, :numeral)
  end

  # Convert an Integer into Sino-Japanese number
  # spelling in hiragana or katakana
  #
  # Examples:
  #   Bangou.integer_to_japanese_text(823)
  #   # => "はっぴゃくにじゅうさん"
  #   Bangou.integer_to_japanese_text(0)
  #   # => "ゼロ"
  #
  # @param [Integer] the number to convert
  # @return [String] The Sino-Japanese numeric equivalent in kana
  def self.integer_to_japanese_text int
    convert_number(int, :text)
  end

  # Detect whether a given number can be processed
  # by this class
  #
  # @return [Boolean]
  def self.in_range? int
    int >= 0 and int < 100000000
  end

  private

  NUMERAL_INDEX = 0
  TEXT_INDEX    = 1
  RANGE_ERROR_MESSAGE = "Bangou can only process numbers between 0 and 99,999,999"

  NUMERALS = {
		0 => ["",""],      1 => ["一","いち"], 2 => ["二","に"],
		3 => ["三","さん"], 4 => ["四","よん"], 5 => ["五","ご"],
		6 => ["六","ろく"], 7 => ["七","なな"], 8 => ["八","はち"],
		9 => ["九","きゅう"]
	}

  BASES = {
    1     => ["",""],
    10    => ["十","じゅう"],
    100   => ["百","ひゃく"],
    1000  => ["千","せん"],
    10000 => ["万","まん"]
  }

  EXCEPTIONS = {
    300  => [nil,"さんびゃく"],
    600  => [nil,"ろっぴゃく"],
    800  => [nil,"はっぴゃく"],
    3000 => [nil,"さんぜん"],
    8000 => [nil,"はっせん"]
  }

  ZEROES = {
    :text    => "ゼロ",
    :numeral => "０"
  }

  def self.index_for_format format
    format == :text ? TEXT_INDEX : NUMERAL_INDEX
  end

  def self.exceptions num, base, format
    hash_index = num * base
    EXCEPTIONS[hash_index][index_for_format(format)] if EXCEPTIONS[hash_index]
  end

  def self.numerals num, format
    NUMERALS[num][index_for_format(format)]
  end

  def self.bases num, format
    BASES[num][index_for_format(format)]
  end

  def self.convert_number int, format
    raise OutOfRangeException.new(RANGE_ERROR_MESSAGE) unless in_range?(int)
    return ZEROES[format] if int == 0

    digits = int.to_s.split(//).reverse

    digits.each_with_index.map do |digit, power|
      hide_base_from_value = has_more_digits_after_index?(digits, power)
      characters_for_digit_in_base(digit.to_i, 10 ** power, format, hide_base_from_value)
    end.reverse.join
  end

  def self.strip_leading_ones text
    text.sub(/^(一|いち)(.+)/,'\2')
  end

  def self.has_more_digits_after_index?(digits, index)
    digits.size > 1 and (digits[0..index - 1] & ("1".."9").to_a).size > 0
  end

  def self.characters_for_digit_in_base(num, base, format, hide_base_from_value)
    return "" if num == 0

    if text = exceptions(num, base, format)
      text
    elsif base <= 10000
      text = numerals(num, format) + bases(base, format)
    else
      text  = characters_for_digit_in_base(num, base/10000, format, hide_base_from_value)
      text += bases(10000, format) unless hide_base_from_value
    end
    strip_leading_ones(text)
  end
end